vendor/engines/bulk_email/app/controllers/bulk_email/bulk_email_controller.rb

```diff
@@ -45,14 +45,33 @@ def deliver
         flash[:notice] = text("bulk_email.delivery.success", count: @delivery_form.recipient_ids.count)
         redirect_to delivery_success_path
       else
-        flash[:error] = text("bulk_email.delivery.failure")
+        flash.now[:error] = text("bulk_email.delivery.failure")
         @users = User.where_ids_in(@delivery_form.recipient_ids)
+        populate_flow_state_params
         render :create
       end
     end
 
     private
 
+    # There are several hidden parameters in the bulk email compose form that
+    # hold state on how the user got here, what product(s) the mail is about,
+    # and how the recipients were chosen. If there is a form error, these params
+    # need to be set to keep that state in the redisplayed form.
+    def populate_flow_state_params
+      return if params[:bulk_email_delivery_form].blank?
+      search_criteria = JSON.parse(params[:bulk_email_delivery_form][:search_criteria])
```

Comment:
You're right on the `||=`; they don't need to be there.

The problem I run into with `bulk_email_recipient_search_params` is that in this context (when form validation fails), those values only exist in the `search_criteria` as JSON. They could conceivably exist as added hidden form fields, though that feels like it would clutter the form with extra params that the "happy path" would just throw away.